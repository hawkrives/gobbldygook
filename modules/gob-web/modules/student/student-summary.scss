.student-summary {
	text-align: center;
	hyphens: none;
	font-variant-numeric: oldstyle-nums tabular-nums;
	font-size: 0.85em;
	padding: calc(var(--block-edge-padding) * 2);

	&.cannot-graduate .graduation-message {
		body:not(.dark) & {
			color: var(--deep-orange-500);
		}
	}

	&.can-graduate .graduation-message {
		body:not(.dark) & {
			color: var(--light-green-800);
		}
	}

	&.can-graduate {
		body:not(.dark) & {
			background-color: var(--light-green-50);
		}
		&::selection {
			body:not(.dark) & {
				background-color: var(--light-green-900);
			}
		}
	}

	.paragraph {
		margin: 0;
	}

	.paragraph + .paragraph {
		margin-top: calc(var(--block-edge-padding) * 2);
	}

	.student-summary--header {
		color: var(--black);
		body.dark & {
			color: var(--inherit);
		}

		margin-bottom: calc(var(--block-edge-padding) * 2);
		font-style: italic;
		line-height: 1.6em;
	}
}

.student-summary--editor {
	display: grid;
	grid-template-columns: max-content 1fr;
	align-items: center;
	grid-column-gap: 0.5em;
	grid-row-gap: 0.5em;

	padding-left: 0.5em;
	padding-right: 0.25em;

	label {
		text-align: right;
	}

	input {
		width: 100%;
		border-radius: 0.2em;
		padding: 2px 6px;
		border: solid 1px var(--gray-300);

		&:focus {
			box-shadow: var(--blue-500) 0 0 0 1px;
			border-color: var(--blue-500);
			outline: 0;
		}
	}

	margin-bottom: 1em;
}

.student-letter {
	display: block;
	margin: var(--block-edge-padding) auto calc(var(--block-edge-padding) * 2);

	border: solid 1px;
	background-color: white;

	&.cannot-graduate {
		color: var(--deep-orange-500);
	}

	&.can-graduate {
		color: var(--light-green-800);
		box-shadow: 0 0 0 5px white, 0 0 0 6px var(--light-green-100);
	}
}
