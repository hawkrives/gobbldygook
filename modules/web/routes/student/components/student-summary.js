// @flow
import React from 'react'
import cx from 'classnames'
import oxford from 'listify'
import plur from 'plur'
import filter from 'lodash/filter'
import map from 'lodash/map'
import sample from 'lodash/sample'

import AvatarLetter from 'modules/web/components/avatar-letter'
import ContentEditable from 'modules/web/components/content-editable'

import { getActiveStudentCourses } from 'modules/web/helpers/get-active-student-courses'
import { countCredits } from 'modules/core/examine-student'

import * as colors from 'modules/web/styles/colors'
import * as variables from 'modules/web/styles/variables'
import * as mixins from 'modules/web/styles/mixins'

const goodGraduationMessage =
	`It looks like you'll make it! Just follow the plan, and go over my output
	with your advisor a few times.`.split('\n').join(' ')

const badGraduationMessage =
	`You haven't planned everything out yet. Ask your advisor if you need help
	fitting everything in.`.split('\n').join(' ')

const welcomeMessages = [
	'Hi, ',
	'Hi there, ',
	'Hello, ',
	'こんにちは、', // japanese
	'ようこそ、', // japanese
	'Fram! Fram! ',
	'Salut, ',
	'Aloha, ',
	'Привет, ',
	'Вітаю, ',
	'Sawubona, ',
	'Hei, ',
	'Hola, ', // spanish
	'Bonjour, ', // french
	'Hallo, ', // german
	'nyob zoo ', // hmong
	'你好，', // mandarin
	'안녕하세요 ', // korean
	'สวัสดี ', // thai
	'halo, ', // indonesian
]

const welcomeMessage = sample(welcomeMessages)

export default function StudentSummary({
	student,
	showMessage=true,
	showAvatar=true,
	randomizeHello=false,
	onChangeName,
	onChangeGraduation,
	onChangeMatriculation,
}: {
	student: Object,
	showMessage?: boolean,
	showAvatar?: boolean,
	randomizeHello?: boolean,
	onChangeName?: (value: string) => any,
	onChangeGraduation?: (value: string) => any,
	onChangeMatriculation?: (value: string) => any,
}) {
	const { studies, canGraduate } = student

	const NameEl = (onChangeName
		? <ContentEditable
			className="autosize-input"
			onBlur={onChangeName}
			value={String(student.name)}
		/>
		: <span>{String(student.name)}</span>)

	const degrees = filter(studies, { type: 'degree' })
	const majors = filter(studies, { type: 'major' })
	const concentrations = filter(studies, { type: 'concentration' })
	const emphases = filter(studies, { type: 'emphasis' })

	const degreeWord = plur('degree', degrees.length)
	const majorWord = plur('major', majors.length)
	const concentrationWord = plur('concentration', concentrations.length)
	const emphasisWord = plur('emphasis', emphases.length)

	const degreeEmphasizer = (degrees.length === 1) ? 'a ' : ''
	const majorEmphasizer = (majors.length === 1) ? 'a ' : ''
	const concentrationEmphasizer = (concentrations.length === 1) ? 'a ' : ''
	const emphasisEmphasizer = (emphases.length === 1) ? 'an ' : ''

	const degreeList = oxford(map(degrees, d => d.name))
	const majorList = oxford(map(majors, m => m.name))
	const concentrationList = oxford(map(concentrations, c => c.name))
	const emphasisList = oxford(map(emphases, e => e.name))

	const currentCredits = countCredits(getActiveStudentCourses(student))
	const neededCredits = student.creditsNeeded
	const enoughCredits = currentCredits >= neededCredits

	const graduationEl = (onChangeGraduation
		? <ContentEditable
			className="autosize-input"
			onBlur={onChangeGraduation}
			value={String(student.graduation)}
		/>
		: <span>{String(student.graduation)}</span>
	)

	const matriculationEl = (onChangeMatriculation
		? <ContentEditable
			className="autosize-input"
			onBlur={onChangeMatriculation}
			value={String(student.matriculation)}
		/>
		: <span>{String(student.matriculation)}</span>
	)

	const canGraduateClass = canGraduate ? 'can-graduate' : 'cannot-graduate'

	return (
		<article className={cx('student-summary', canGraduateClass)}>
			<header>
				{showAvatar ? <AvatarLetter
					className={cx('student-letter', canGraduateClass)}
					value={student.name}
				/> : null}
				<div className="intro">
					{randomizeHello ? sample(welcomeMessages) : welcomeMessage}{NameEl}!
				</div>
			</header>
			<div className="content">
				<p>
					After matriculating in {matriculationEl}, you are planning to graduate in {graduationEl}, with {' '}
					{(degrees.length > 0) ? `${degreeEmphasizer}${degreeList} ${degreeWord}` : `no ${degreeWord}`}
					{(majors.length || concentrations.length || emphases.length) ? (majors.length) && (concentrations.length || emphases.length) ? ', ' : ' and ' : ''}
					{(majors.length > 0) && `${majorEmphasizer}${majorWord} in ${majorList}`}
					{(majors.length && concentrations.length) ? ', and ' : ''}
					{(concentrations.length > 0) && `${concentrationEmphasizer}${concentrationWord} in ${concentrationList}`}
					{((majors.length || concentrations.length) && emphases.length) ? ', ' : ''}
					{(emphases.length > 0) && `not to mention ${emphasisEmphasizer}${emphasisWord} in ${emphasisList}`}
					{'. '}
					{currentCredits ? `You have currently planned for ${currentCredits} of your ${neededCredits} required credits. ${enoughCredits ? 'Good job!' : ''}`: ''}
				</p>
				{showMessage ? <p className="graduation-message">
					{canGraduate ? goodGraduationMessage : badGraduationMessage}
				</p>: null}
			</div>
			<style jsx>{`
				.student-summary {
					${mixins.card}

					text-align: center;
					hyphens: auto;
					font-feature-settings: 'onum', 'tnum';
					text-shadow: 0 1px ${colors.white_transparent_50};
				}

				.student-summary.cannot-graduate .graduation-message {
					color: ${colors.deep_orange_500};
				}

				.student-summary.can-graduate .graduation-message {
					color: ${colors.light_green_800};
				}
				.student-summary.can-graduate {
					background-color: ${colors.light_green_50};
				}
				.student-summary.can-graduate::selection {
					background-color: ${colors.light_green_900};
				}

				.content {
					padding: calc(${variables.block_edge_padding} * 2);
				}

				.content p {
					margin: 0;
				}

				.content p + p {
					margin-top: calc(${variables.block_edge_padding} * 2);
				}

				header {
					color: ${colors.black};

					margin: calc(${variables.block_edge_padding} * 2);
					margin-bottom: 0;
					font-style: italic;
					line-height: 1.6em;
				}

				:global(.autosize-input) {
					border-bottom: 1px solid ${colors.gray_400};
					padding: 0 0.125em 0 0;
				}

				:global(.autosize-input):focus {
					outline: 0;
					background-color: ${colors.blue_50};
					border-color: ${colors.blue_500};
				}

				.student-letter {
					display: block;
					margin: ${variables.block_edge_padding} auto calc(${variables.block_edge_padding} * 2);

					border: solid 1px;
					background-color: ${colors.white};
				}

				.student-letter.cannot-graduate {
					color: ${colors.deep_orange_500};
				}

				.student-letter.can-graduate {
					color: ${colors.light_green_800};
					box-shadow:
						0 0 0 5px ${colors.white},
						0 0 0 6px ${colors.light_green_100};
				}
			`}</style>
		</article>
	)
}
