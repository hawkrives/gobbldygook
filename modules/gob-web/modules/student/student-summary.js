// @flow
import React from 'react'
import cx from 'classnames'
import listify from 'listify'
import groupBy from 'lodash/groupBy'
import sample from 'lodash/sample'

import {AvatarLetter} from '../../components/avatar-letter'
import ContentEditable from '../../components/content-editable'

import {getActiveStudentCourses} from '../../helpers/get-active-student-courses'
import {countCredits} from '@gob/examine-student'

import './student-summary.scss'

const welcomeMessages = [
	'Hi, ',
	'Hi there, ',
	'Hello, ',
	'こんにちは、', // japanese
	'ようこそ、', // japanese
	'Fram! Fram! ',
	'Salut, ',
	'Aloha, ', // hawaiian
	'Привет, ',
	'Вітаю, ',
	'Sawubona, ',
	'Hei, ',
	'Hola, ', // spanish
	'Bonjour, ', // french
	'Hallo, ', // german
	'nyob zoo ', // hmong
	'你好，', // mandarin
	'안녕하세요 ', // korean
	'สวัสดี ', // thai
	'halo, ', // indonesian
	'Salve, ', // latin
	'Χαῖρε! ', // ancient greek
]

const welcomeMessage = welcomeMessages[2]

type Student = Object

type Props = {
	onChangeGraduation?: string => any,
	onChangeMatriculation?: string => any,
	onChangeName?: string => any,
	randomizeHello?: boolean,
	showAvatar?: boolean,
	showMessage?: boolean,
	student: Student,
}

export class StudentSummary extends React.PureComponent<Props> {
	render() {
		const {
			student,
			showMessage = true,
			showAvatar = true,
			randomizeHello = true,
			...props
		} = this.props
		const {studies, canGraduate} = student

		const className = canGraduate ? 'can-graduate' : 'cannot-graduate'

		const currentCredits = countCredits(getActiveStudentCourses(student))
		const neededCredits = student.creditsNeeded

		const message = randomizeHello
			? sample(welcomeMessages)
			: welcomeMessage

		return (
			<article className={cx('student-summary', className)}>
				<Header
					canGraduate={canGraduate}
					name={student.name}
					onChangeName={props.onChangeName}
					helloMessage={message}
					showAvatar={showAvatar}
				/>
				<div className="content">
					<DateSummary
						onChangeGraduation={props.onChangeGraduation}
						onChangeMatriculation={props.onChangeMatriculation}
						matriculation={student.matriculation}
						graduation={student.graduation}
					/>

					<DegreeSummary studies={studies} />

					<CreditSummary
						currentCredits={currentCredits}
						neededCredits={neededCredits}
					/>

					{showMessage ? <Footer canGraduate={canGraduate} /> : null}
				</div>
			</article>
		)
	}
}

type HeaderProps = {
	canGraduate: boolean,
	helloMessage: string,
	name: string,
	onChangeName?: string => any,
	showAvatar: boolean,
}

export class Header extends React.PureComponent<HeaderProps> {
	render() {
		const props = this.props

		const className = props.canGraduate ? 'can-graduate' : 'cannot-graduate'

		const avatar = props.showAvatar ? (
			<AvatarLetter
				className={cx('student-letter', className)}
				value={props.name}
			/>
		) : null

		const name = (
			<ContentEditable
				disabled={!props.onChangeName}
				className="autosize-input"
				onBlur={props.onChangeName}
				value={String(props.name)}
			/>
		)

		return (
			<header className="student-summary--header">
				{avatar}

				<div className="intro">
					{props.helloMessage}
					{name}!
				</div>
			</header>
		)
	}
}

type FooterProps = {
	canGraduate: boolean,
}

export class Footer extends React.PureComponent<FooterProps> {
	goodGraduationMessage =
		"It looks like you'll make it! Just follow the plan, and go over my output with your advisor a few times."

	badGraduationMessage =
		"You haven't planned everything out yet. Ask your advisor if you need help fitting everything in."

	render() {
		const msg = this.props.canGraduate
			? this.goodGraduationMessage
			: this.badGraduationMessage

		return <div className="paragraph graduation-message">{msg}</div>
	}
}

type DateSummaryProps = {
	onChangeGraduation?: string => any,
	onChangeMatriculation?: string => any,
	matriculation: string,
	graduation: string,
}

export class DateSummary extends React.PureComponent<DateSummaryProps> {
	render() {
		const props = this.props

		const matriculation = (
			<ContentEditable
				disabled={!props.onChangeMatriculation}
				className="autosize-input"
				onBlur={props.onChangeMatriculation}
				value={String(props.matriculation)}
			/>
		)

		const graduation = (
			<ContentEditable
				disabled={!props.onChangeGraduation}
				className="autosize-input"
				onBlur={props.onChangeGraduation}
				value={String(props.graduation)}
			/>
		)

		return (
			<p className="paragraph">
				After matriculating in {matriculation}, you are planning to
				graduate in {graduation}.
			</p>
		)
	}
}

type DegreeSummaryProps = {
	studies: {type: string, name: string}[],
}

export class DegreeSummary extends React.PureComponent<DegreeSummaryProps> {
	render() {
		const grouped: {
			[key: string]: {type: string, name: string}[],
		} = groupBy(this.props.studies, s => s.type)

		const {
			degree: dS = [],
			major: mS = [],
			concentration: cS = [],
			emphasis: eS = [],
		} = grouped

		const dCount = dS.length
		const mCount = mS.length
		const cCount = cS.length
		const eCount = eS.length

		const dWord = dCount === 1 ? 'degree' : 'degrees'
		const mWord = mCount === 1 ? 'major' : 'majors'
		const cWord = cCount === 1 ? 'concentration' : 'concentrations'
		const eWord = eCount === 1 ? 'emphasis' : 'emphases'

		const dEmph = dCount === 1 ? 'a ' : ''
		const mEmph = mCount === 1 ? 'a ' : ''
		const cEmph = cCount === 1 ? 'a ' : ''
		const eEmph = eCount === 1 ? 'an ' : ''

		const dList = listify(dS.map(d => d.name))
		const mList = listify(mS.map(m => m.name))
		const cList = listify(cS.map(c => c.name))
		const eList = listify(eS.map(e => e.name))

		return (
			<p className="paragraph">
				You are planning on{' '}
				{dCount > 0 ? `${dEmph}${dList} ${dWord}` : `no ${dWord}`}
				{mCount || cCount || eCount
					? mCount && (cCount || eCount)
						? ', '
						: ' and '
					: ''}
				{mCount ? `${mEmph}${mWord} in ${mList}` : ''}
				{mCount && cCount ? ', and ' : ''}
				{cCount ? `${cEmph}${cWord} in ${cList}` : ''}
				{(mCount || cCount) && eCount ? ', ' : ''}
				{eCount ? `not to mention ${eEmph}${eWord} in ${eList}` : ''}
				{'.'}
			</p>
		)
	}
}

type CreditSummaryProps = {
	currentCredits: number,
	neededCredits: number,
}

export class CreditSummary extends React.PureComponent<CreditSummaryProps> {
	render() {
		const {currentCredits, neededCredits} = this.props
		const enoughCredits = currentCredits >= neededCredits

		return (
			<p className="paragraph">
				You have currently planned for {currentCredits} of your{' '}
				{neededCredits} required credits.
				{enoughCredits ? ' Good job!' : ''}
			</p>
		)
	}
}
