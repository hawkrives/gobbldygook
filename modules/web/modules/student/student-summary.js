// @flow
import React from 'react'
import cx from 'classnames'
import listify from 'listify'
import groupBy from 'lodash/groupBy'
import sample from 'lodash/sample'

import AvatarLetter from '../../components/avatar-letter'
import ContentEditable from '../../components/content-editable'

import {
    getActiveStudentCourses,
} from '../../helpers/get-active-student-courses'
import { countCredits } from '../../../examine-student/count-credits'

import './student-summary.scss'

const welcomeMessages = [
    'Hi, ',
    'Hi there, ',
    'Hello, ',
    'こんにちは、', // japanese
    'ようこそ、', // japanese
    'Fram! Fram! ',
    'Salut, ',
    'Aloha, ',
    'Привет, ',
    'Вітаю, ',
    'Sawubona, ',
    'Hei, ',
    'Hola, ', // spanish
    'Bonjour, ', // french
    'Hallo, ', // german
    'nyob zoo ', // hmong
    '你好，', // mandarin
    '안녕하세요 ', // korean
    'สวัสดี ', // thai
    'halo, ', // indonesian
]

const welcomeMessage = welcomeMessages[2]

type Student = Object;

export class StudentSummary extends React.PureComponent {
    props: {
        onChangeGraduation?: string => any,
        onChangeMatriculation?: string => any,
        onChangeName?: string => any,
        randomizeHello?: boolean,
        showAvatar?: boolean,
        showMessage?: boolean,
        student: Student,
    };

    render() {
        const {
            student,
            showMessage = true,
            showAvatar = true,
            randomizeHello = true,
            ...props
        } = this.props
        const { studies, canGraduate } = student

        const className = canGraduate ? 'can-graduate' : 'cannot-graduate'

        const currentCredits = countCredits(getActiveStudentCourses(student))
        const neededCredits = student.creditsNeeded

        const message = randomizeHello
            ? sample(welcomeMessages)
            : welcomeMessage

        return (
            <article className={cx('student-summary', className)}>
                <Header
                    canGraduate={canGraduate}
                    name={student.name}
                    onChangeName={props.onChangeName}
                    helloMessage={message}
                    showAvatar={showAvatar}
                />
                <div className="content">
                    <DateSummary
                        onChangeGraduation={props.onChangeGraduation}
                        onChangeMatriculation={props.onChangeMatriculation}
                        matriculation={student.matriculation}
                        graduation={student.graduation}
                    />

                    <DegreeSummary studies={studies} />

                    <CreditSummary
                        currentCredits={currentCredits}
                        neededCredits={neededCredits}
                    />

                    {showMessage ? <Footer canGraduate={canGraduate} /> : null}
                </div>
            </article>
        )
    }
}

export class Header extends React.PureComponent {
    props: {
        canGraduate: boolean,
        helloMessage: string,
        name: string,
        onChangeName?: string => any,
        showAvatar: boolean,
    };

    render() {
        const props = this.props

        const className = props.canGraduate
            ? 'can-graduate'
            : 'cannot-graduate'

        const avatar = props.showAvatar
            ? <AvatarLetter
                  className={cx('student-letter', className)}
                  value={props.name}
              />
            : null

        const name = (
            <ContentEditable
                disabled={!props.onChangeName}
                className="autosize-input"
                onBlur={props.onChangeName}
                value={String(props.name)}
            />
        )

        return (
            <header className="student-summary--header">
                {avatar}

                <div className="intro">
                    {props.helloMessage}{name}!
                </div>
            </header>
        )
    }
}

export class Footer extends React.PureComponent {
    goodGraduationMessage = [
        "It looks like you'll make it! Just follow the plan, and",
        'go over my output with your advisor a few times.',
    ].join(' ');

    badGraduationMessage = [
        "You haven't planned everything out yet.",
        'Ask your advisor if you need help fitting everything in.',
    ].join(' ');

    props: {
        canGraduate: boolean,
    };

    render() {
        const msg = this.props.canGraduate
            ? this.goodGraduationMessage
            : this.badGraduationMessage

        return (
            <div className="paragraph graduation-message">
                {msg}
            </div>
        )
    }
}

export class DateSummary extends React.PureComponent {
    props: {
        onChangeGraduation?: string => any,
        onChangeMatriculation?: string => any,
        matriculation: string,
        graduation: string,
    };

    render() {
        const props = this.props

        const matriculation = (
            <ContentEditable
                disabled={!props.onChangeMatriculation}
                className="autosize-input"
                onBlur={props.onChangeMatriculation}
                value={String(props.matriculation)}
            />
        )

        const graduation = (
            <ContentEditable
                disabled={!props.onChangeGraduation}
                className="autosize-input"
                onBlur={props.onChangeGraduation}
                value={String(props.graduation)}
            />
        )

        return (
            <p className="paragraph">
                After matriculating in {matriculation}, you are
                planning to graduate in {graduation}.
            </p>
        )
    }
}

export class DegreeSummary extends React.PureComponent {
    props: {
        studies: { type: string, name: string }[],
    };

    render() {
        const grouped: {
            [key: string]: { type: string, name: string }[],
        } = groupBy(this.props.studies, s => s.type)

        const {
            degree: dS = [],
            major: mS = [],
            concentration: cS = [],
            emphasis: eS = [],
        } = grouped

        const dCount = dS.length
        const mCount = mS.length
        const cCount = cS.length
        const eCount = eS.length

        const dWord = dCount === 1 ? 'degree' : 'degrees'
        const mWord = mCount === 1 ? 'major' : 'majors'
        const cWord = cCount === 1 ? 'concentration' : 'concentrations'
        const eWord = eCount === 1 ? 'emphasis' : 'emphases'

        const dEmph = dCount === 1 ? 'a ' : ''
        const mEmph = mCount === 1 ? 'a ' : ''
        const cEmph = cCount === 1 ? 'a ' : ''
        const eEmph = eCount === 1 ? 'an ' : ''

        const dList = listify(dS.map(d => d.name))
        const mList = listify(mS.map(m => m.name))
        const cList = listify(cS.map(c => c.name))
        const eList = listify(eS.map(e => e.name))

        return (
            <p className="paragraph">
                You are planning on{' '}
                {dCount > 0 ? `${dEmph}${dList} ${dWord}` : `no ${dWord}`}
                {mCount || cCount || eCount
                    ? mCount && (cCount || eCount) ? ', ' : ' and '
                    : ''}
                {mCount ? `${mEmph}${mWord} in ${mList}` : ''}
                {mCount && cCount ? ', and ' : ''}
                {cCount ? `${cEmph}${cWord} in ${cList}` : ''}
                {(mCount || cCount) && eCount ? ', ' : ''}
                {eCount ? `not to mention ${eEmph}${eWord} in ${eList}` : ''}
                {'.'}
            </p>
        )
    }
}

export class CreditSummary extends React.PureComponent {
    props: {
        currentCredits: number,
        neededCredits: number,
    };

    render() {
        const { currentCredits, neededCredits } = this.props
        const enoughCredits = currentCredits >= neededCredits

        return (
            <p className="paragraph">
                You have currently planned for {currentCredits}{' '}
                of your {neededCredits} required credits.
                {enoughCredits ? ' Good job!' : ''}
            </p>
        )
    }
}
