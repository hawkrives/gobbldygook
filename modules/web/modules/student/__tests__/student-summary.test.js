// @flow

import React from 'react'
import { CreditSummary, DateSummary } from '../student-summary'
import { shallow } from 'enzyme'

describe('CreditSummary', () => {
    it('renders shallowly', () => {
        const tree = shallow(
            <CreditSummary currentCredits={5} neededCredits={10} />
        )

        expect(tree).toMatchSnapshot()
    })

    it('handles having fewer credits than needed', () => {
        const tree = shallow(
            <CreditSummary currentCredits={5} neededCredits={10} />
        )

        expect(tree).toMatchSnapshot()
        expect(tree.text()).not.toContain('Good job!')
    })

    it('handles having exactly the right number of credits', () => {
        const tree = shallow(
            <CreditSummary currentCredits={10} neededCredits={10} />
        )

        expect(tree).toMatchSnapshot()
        expect(tree.text()).toContain('Good job!')
    })

    it('handles having more credits than needed', () => {
        const tree = shallow(
            <CreditSummary currentCredits={15} neededCredits={10} />
        )

        expect(tree).toMatchSnapshot()
        expect(tree.text()).toContain('Good job!')
    })
})

describe('DateSummary', () => {
    it('renders', () => {
        const tree = shallow(
            <DateSummary matriculation="2012" graduation="2016" />
        )

        expect(tree).toMatchSnapshot()
    })

    it('handles graduating before matriculation', () => {
        expect(
            shallow(<DateSummary matriculation="2016" graduation="2012" />)
        ).toMatchSnapshot()
    })

    it('handles graduating in three years', () => {
        expect(
            shallow(<DateSummary matriculation="2000" graduation="2003" />)
        ).toMatchSnapshot()
    })

    it('handles graduating in four years', () => {
        expect(
            shallow(<DateSummary matriculation="2000" graduation="2004" />)
        ).toMatchSnapshot()
    })

    it('handles graduating in five years', () => {
        expect(
            shallow(<DateSummary matriculation="2000" graduation="2005" />)
        ).toMatchSnapshot()
    })

    it('handles graduating in six years', () => {
        expect(
            shallow(<DateSummary matriculation="2000" graduation="2006" />)
        ).toMatchSnapshot()
    })

    it('disables editing the matriculation year if onChangeMatriculation is not given', () => {
        const tree = shallow(
            <DateSummary matriculation="2012" graduation="2016" />
        )

        expect(tree).toMatchSnapshot()
        expect(tree.find('ContentEditable').at(0).prop('disabled')).toBe(true)
    })

    it('disables editing the graduation year if onChangeGraduation is not given', () => {
        const tree = shallow(
            <DateSummary matriculation="2012" graduation="2016" />
        )

        expect(tree).toMatchSnapshot()
        expect(tree.find('ContentEditable').at(1).prop('disabled')).toBe(true)
    })

    it('allows editing the matriculation year if onChangeMatriculation is not given', () => {
        const onChangeMatriculation = jest.fn()
        const tree = shallow(
            <DateSummary
                onChangeMatriculation={onChangeMatriculation}
                matriculation="2012"
                graduation="2016"
            />
        )

        expect(tree).toMatchSnapshot()
        expect(tree.find('ContentEditable').at(0).prop('disabled')).toBe(false)
    })

    it('allows editing the graduation year if onChangeGraduation is not given', () => {
        const onChangeGraduation = jest.fn()
        const tree = shallow(
            <DateSummary
                onChangeGraduation={onChangeGraduation}
                matriculation="2012"
                graduation="2016"
            />
        )

        expect(tree).toMatchSnapshot()
        expect(tree.find('ContentEditable').at(1).prop('disabled')).toBe(false)
    })

    it('calls onChangeMatriculation when the matriculation year is changed', () => {
        const onChangeMatriculation = jest.fn()
        const tree = shallow(
            <DateSummary
                onChangeMatriculation={onChangeMatriculation}
                matriculation="2012"
                graduation="2016"
            />
        )

        expect(tree).toMatchSnapshot()
        tree.find('ContentEditable').at(0).simulate('blur', '2010')
        expect(onChangeMatriculation).toHaveBeenCalledWith('2010')
    })

    it('calls onChangeGraduation when the graduation year is changed', () => {
        const onChangeGraduation = jest.fn()
        const tree = shallow(
            <DateSummary
                onChangeGraduation={onChangeGraduation}
                matriculation="2012"
                graduation="2016"
            />
        )

        expect(tree).toMatchSnapshot()
        tree.find('ContentEditable').at(1).simulate('blur', '2018')
        expect(onChangeGraduation).toHaveBeenCalledWith('2018')
    })
})
