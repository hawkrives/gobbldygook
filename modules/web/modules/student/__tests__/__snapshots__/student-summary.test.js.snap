// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditSummary handles having exactly the right number of credits 1`] = `
<p
  className="paragraph"
>
  You have currently planned for 
  10
   
  of your 
  10
   required credits.
   Good job!
</p>
`;

exports[`CreditSummary handles having fewer credits than needed 1`] = `
<p
  className="paragraph"
>
  You have currently planned for 
  5
   
  of your 
  10
   required credits.
</p>
`;

exports[`CreditSummary handles having more credits than needed 1`] = `
<p
  className="paragraph"
>
  You have currently planned for 
  15
   
  of your 
  10
   required credits.
   Good job!
</p>
`;

exports[`CreditSummary renders shallowly 1`] = `
<p
  className="paragraph"
>
  You have currently planned for 
  5
   
  of your 
  10
   required credits.
</p>
`;

exports[`DateSummary allows editing the graduation year if onChangeGraduation is not given 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2012"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={false}
    multiLine={false}
    onBlur={[Function]}
    onChange={[Function]}
    value="2016"
  />
  .
</p>
`;

exports[`DateSummary allows editing the matriculation year if onChangeMatriculation is not given 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={false}
    multiLine={false}
    onBlur={[Function]}
    onChange={[Function]}
    value="2012"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2016"
  />
  .
</p>
`;

exports[`DateSummary calls onChangeGraduation when the graduation year is changed 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2012"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={false}
    multiLine={false}
    onBlur={[Function]}
    onChange={[Function]}
    value="2016"
  />
  .
</p>
`;

exports[`DateSummary calls onChangeMatriculation when the matriculation year is changed 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={false}
    multiLine={false}
    onBlur={[Function]}
    onChange={[Function]}
    value="2012"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2016"
  />
  .
</p>
`;

exports[`DateSummary disables editing the graduation year if onChangeGraduation is not given 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2012"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2016"
  />
  .
</p>
`;

exports[`DateSummary disables editing the matriculation year if onChangeMatriculation is not given 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2012"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2016"
  />
  .
</p>
`;

exports[`DateSummary handles graduating before matriculation 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2016"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2012"
  />
  .
</p>
`;

exports[`DateSummary handles graduating in five years 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2000"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2005"
  />
  .
</p>
`;

exports[`DateSummary handles graduating in four years 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2000"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2004"
  />
  .
</p>
`;

exports[`DateSummary handles graduating in six years 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2000"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2006"
  />
  .
</p>
`;

exports[`DateSummary handles graduating in three years 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2000"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2003"
  />
  .
</p>
`;

exports[`DateSummary renders 1`] = `
<p
  className="paragraph"
>
  After matriculating in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2012"
  />
  , you are planning to graduate in 
  <ContentEditable
    className="autosize-input"
    disabled={true}
    multiLine={false}
    onChange={[Function]}
    value="2016"
  />
  .
</p>
`;
